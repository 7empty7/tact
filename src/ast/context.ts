import { parse } from "../grammar/grammar";
import { ASTNode, ASTType } from "./ast";

export type VariableRef = {
    name: string,
    node: ASTNode
};

export class CompilerContext {

    static fromSources(sources: string[]) {
        let asts = sources.map(source => parse(source));
        let ctx = new CompilerContext();
        for (let a of asts) {
            for (let e of a.entries) {
                if (e.kind === 'def_struct') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'def_contract') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'primitive') {
                    ctx = ctx.addASTType(e);
                }
            }
        }
        return ctx;
    }

    readonly astTypes: { [key: string]: ASTType };
    readonly variables: { [key: string]: VariableRef };
    readonly shared: { [key: symbol]: any } = {};

    constructor(astTypes: { [key: string]: ASTType } = {}, variables: { [key: string]: VariableRef } = {}, shared: { [key: symbol]: any } = {}) {
        this.astTypes = astTypes;
        this.variables = variables;
        this.shared = shared;
        Object.freeze(this.astTypes);
        Object.freeze(this.shared);
        Object.freeze(this.variables);
        Object.freeze(this);
    }

    addASTType = (ref: ASTType) => {
        if (this.astTypes[ref.name]) {
            throw Error('Type already exists');
        }
        return new CompilerContext({ ...this.astTypes, [ref.name]: ref }, this.variables, this.shared);
    }

    addVariable = (ref: VariableRef) => {
        if (this.variables[ref.name]) {
            throw Error('Variable already exists');
        }
        return new CompilerContext(this.astTypes, { ...this.variables, [ref.name]: ref }, this.shared);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext(this.astTypes, this.variables, { ...this.shared, [store]: sh });
    }
}

export function createContextStore<T>() {
    let symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            let m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}