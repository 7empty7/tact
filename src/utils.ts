// // export type Interval = 
// export function thr