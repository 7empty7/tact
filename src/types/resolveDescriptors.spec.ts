import { CompilerContext } from "../context";
import { getAllStaticFunctions, getAllTypes, resolveDescriptors } from "./resolveDescriptors";
import { ASTRef, __DANGER_resetNodeId } from "../grammar/ast";
import { loadCases } from "../utils/loadCases";
import { openContext } from "../grammar/store";

expect.addSnapshotSerializer({
    test: (src) => src instanceof ASTRef,
    print: (src) => `${(src as ASTRef).contents}`
});

describe('resolveDescriptors', () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    for (let r of loadCases(__dirname + "/test/")) {
        it('should resolve descriptors for ' + r.name, () => {
            let ctx = openContext(new CompilerContext(), [r.code]);
            ctx = resolveDescriptors(ctx);
            expect(getAllTypes(ctx)).toMatchSnapshot();
            expect(getAllStaticFunctions(ctx)).toMatchSnapshot();
        });
    }
    for (let r of loadCases(__dirname + "/test-failed/")) {
        it('should fail descriptors for ' + r.name, () => {
            let ctx = openContext(new CompilerContext(), [r.code]);
            expect(() => resolveDescriptors(ctx)).toThrowErrorMatchingSnapshot();
        });
    }
});