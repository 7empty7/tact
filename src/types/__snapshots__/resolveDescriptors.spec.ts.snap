// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveDescriptors should fail descriptors for case-0 1`] = `
"Line 5, col 1:
  4 | @name(hello_world)
> 5 | get native sample(a: Int): Int;
      ^~~
Native functions cannot be getters"
`;

exports[`resolveDescriptors should fail descriptors for case-1 1`] = `
"Line 5, col 1:
  4 | @name(hello_world)
> 5 | public native sample(a: Int): Int;
      ^~~~~~
Native functions cannot be public"
`;

exports[`resolveDescriptors should fail descriptors for case-2 1`] = `
"Line 4, col 1:
  3 | 
> 4 | public fun sample(a: Int): Int {
      ^~~~~~
  5 |     return 0;
Public functions must be defined within a contract"
`;

exports[`resolveDescriptors should fail descriptors for case-3 1`] = `
"Line 4, col 1:
  3 | 
> 4 | get fun sample(a: Int): Int {
      ^~~
  5 |     return 0;
Getters must be defined within a contract"
`;

exports[`resolveDescriptors should fail descriptors for case-4 1`] = `
"Line 4, col 17:
  3 | 
> 4 | extends fun inc(a: Int): Int {
                      ^~~~~~
  5 |     return a + 1;
Extend function must have first argument named "self""
`;

exports[`resolveDescriptors should fail descriptors for case-5 1`] = `
"Line 4, col 17:
  3 | 
> 4 | fun inc(a: Int, self: Int): Int {
                      ^~~~~~~~~
  5 |     return a + 1 + self;
Argument name "self" is reserved"
`;

exports[`resolveDescriptors should fail descriptors for case-6 1`] = `
"Line 7, col 3:
  6 |   b: Int;
> 7 |   b: Int;
        ^~~~~~~
  8 | }
Field b already exists"
`;

exports[`resolveDescriptors should fail descriptors for case-7 1`] = `
"Line 7, col 3:
  6 |   b: Int;
> 7 |   b: Int;
        ^~~~~~~
  8 | }
Field b already exists"
`;

exports[`resolveDescriptors should fail descriptors for case-8 1`] = `
"Line 9, col 11:
   8 | contract Pair {
>  9 |   receive(src: B?) {
                 ^~~~~~~
  10 |     
Receive function cannot have optional argument"
`;

exports[`resolveDescriptors should fail descriptors for case-9 1`] = `
"Line 9, col 11:
   8 | contract Pair {
>  9 |   receive(src: B) {
                 ^~~~~~
  10 |     
Receive function can only accept message"
`;

exports[`resolveDescriptors should fail descriptors for case-10 1`] = `
"Line 9, col 11:
   8 | contract Pair {
>  9 |   receive(src: Int) {
                 ^~~~~~~~
  10 |     
Receive function can only accept message"
`;

exports[`resolveDescriptors should fail descriptors for case-11 1`] = `
"Line 8, col 23:
  7 |     }
> 8 |     get fun hello2(): Point {
                            ^~~~~
  9 |         return 0;
Type Point not found"
`;

exports[`resolveDescriptors should fail descriptors for case-12 1`] = `
"Line 8, col 1:
  7 | 
> 8 | struct Main {
      ^~~~~~~~~~~~~
  9 |     
Type Main already exists"
`;

exports[`resolveDescriptors should fail descriptors for case-13 1`] = `
"Line 8, col 1:
  7 | 
> 8 | struct Main {
      ^~~~~~~~~~~~~
  9 |     
Type Main already exists"
`;

exports[`resolveDescriptors should fail descriptors for case-14 1`] = `
"Line 8, col 1:
  7 | 
> 8 | contract Main {
      ^~~~~~~~~~~~~~~
  9 |     
Type Main already exists"
`;

exports[`resolveDescriptors should fail descriptors for case-15 1`] = `
"Line 4, col 1:
  3 | 
> 4 | mutates fun hello(self: Main): Int {
      ^~~~~~~
  5 | 
Mutating functions must be extend functions"
`;

exports[`resolveDescriptors should fail descriptors for case-16 1`] = `
"Line 8, col 1:
  7 | 
> 8 | fun hello(b: Main): Int {
      ^~~~~~~~~~~~~~~~~~~~~~~~~
  9 | 
Static function hello already exists"
`;

exports[`resolveDescriptors should fail descriptors for case-17 1`] = `
"Line 4, col 1:
  3 | 
> 4 | contract Main {
      ^~~~~~~~~~~~~~~
  5 |     fun hello(a: Main): Int {
Static function hello already exists in type Main"
`;

exports[`resolveDescriptors should fail descriptors for case-18 1`] = `
"Line 5, col 24:
  4 | contract Main {
> 5 |     fun hello(a: Bool, a: Int): Int {
                             ^~~~~~
  6 | 
Argument name "a" is already used"
`;

exports[`resolveDescriptors should fail descriptors for case-19 1`] = `
"Line 9, col 5:
   8 | 
>  9 |     init(a: Bool) {
           ^~~~~~~~~~~~~~~
  10 | 
Init function already exists"
`;

exports[`resolveDescriptors should fail descriptors for case-20 1`] = `
"Line 8, col 1:
  7 | 
> 8 | extends fun hello(self: Main): Bool {
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  9 |     return true;
Function hello already exists in type Main"
`;

exports[`resolveDescriptors should resolve descriptors for case-0 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-0 2`] = `{}`;

exports[`resolveDescriptors should resolve descriptors for case-1 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
  "Struct1": {
    "ast": {
      "fields": [
        {
          "as": null,
          "id": 4,
          "init": undefined,
          "kind": "def_field",
          "name": "a",
          "ref": a: Int;,
          "type": {
            "id": 3,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
        },
        {
          "as": null,
          "id": 6,
          "init": undefined,
          "kind": "def_field",
          "name": "b",
          "ref": b: Bool;,
          "type": {
            "id": 5,
            "kind": "type_ref_simple",
            "name": "Bool",
            "optional": false,
            "ref": Bool,
          },
        },
      ],
      "id": 7,
      "kind": "def_struct",
      "message": false,
      "name": "Struct1",
      "ref": struct Struct1 {
    a: Int;
    b: Bool;
},
    },
    "fields": [
      {
        "as": null,
        "default": undefined,
        "index": 0,
        "name": "a",
        "type": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
      },
      {
        "as": null,
        "default": undefined,
        "index": 1,
        "name": "b",
        "type": {
          "kind": "ref",
          "name": "Bool",
          "optional": false,
        },
      },
    ],
    "functions": {},
    "init": null,
    "kind": "struct",
    "name": "Struct1",
    "receivers": {},
  },
  "Struct2": {
    "ast": {
      "fields": [
        {
          "as": null,
          "id": 9,
          "init": undefined,
          "kind": "def_field",
          "name": "a",
          "ref": a: Struct1;,
          "type": {
            "id": 8,
            "kind": "type_ref_simple",
            "name": "Struct1",
            "optional": false,
            "ref": Struct1,
          },
        },
      ],
      "id": 10,
      "kind": "def_struct",
      "message": false,
      "name": "Struct2",
      "ref": struct Struct2 {
    a: Struct1;
},
    },
    "fields": [
      {
        "as": null,
        "default": undefined,
        "index": 0,
        "name": "a",
        "type": {
          "kind": "ref",
          "name": "Struct1",
          "optional": false,
        },
      },
    ],
    "functions": {},
    "init": null,
    "kind": "struct",
    "name": "Struct2",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-1 2`] = `{}`;

exports[`resolveDescriptors should resolve descriptors for case-2 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-2 2`] = `
{
  "testFunc": {
    "args": [],
    "ast": {
      "args": [],
      "attributes": [],
      "id": 6,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return 0;
},
      "return": {
        "id": 3,
        "kind": "type_ref_simple",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
      "statements": [
        {
          "expression": {
            "id": 4,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 5,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    "isGetter": false,
    "isMutating": false,
    "isPublic": false,
    "name": "testFunc",
    "returns": {
      "kind": "ref",
      "name": "Int",
      "optional": false,
    },
    "self": null,
  },
  "testFunc2": {
    "args": [],
    "ast": {
      "args": [],
      "attributes": [],
      "id": 9,
      "kind": "def_function",
      "name": "testFunc2",
      "ref": fun testFunc2() {
    return 0;
},
      "return": null,
      "statements": [
        {
          "expression": {
            "id": 7,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 8,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    "isGetter": false,
    "isMutating": false,
    "isPublic": false,
    "name": "testFunc2",
    "returns": {
      "kind": "void",
    },
    "self": null,
  },
  "testFunc3": {
    "args": [
      {
        "as": null,
        "name": "a",
        "ref": a: Int,
        "type": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
      },
      {
        "as": null,
        "name": "b",
        "ref": b: Bool,
        "type": {
          "kind": "ref",
          "name": "Bool",
          "optional": false,
        },
      },
    ],
    "ast": {
      "args": [
        {
          "id": 11,
          "kind": "def_argument",
          "name": "a",
          "ref": a: Int,
          "type": {
            "id": 10,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
        },
        {
          "id": 13,
          "kind": "def_argument",
          "name": "b",
          "ref": b: Bool,
          "type": {
            "id": 12,
            "kind": "type_ref_simple",
            "name": "Bool",
            "optional": false,
            "ref": Bool,
          },
        },
      ],
      "attributes": [],
      "id": 16,
      "kind": "def_function",
      "name": "testFunc3",
      "ref": fun testFunc3(a: Int, b: Bool) {
    return 0;
},
      "return": null,
      "statements": [
        {
          "expression": {
            "id": 14,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 15,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    "isGetter": false,
    "isMutating": false,
    "isPublic": false,
    "name": "testFunc3",
    "returns": {
      "kind": "void",
    },
    "self": null,
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-3 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "HelloWorld": {
    "ast": {
      "declarations": [
        {
          "as": null,
          "id": 4,
          "init": undefined,
          "kind": "def_field",
          "name": "a",
          "ref": a: Int;,
          "type": {
            "id": 3,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
        },
        {
          "as": null,
          "id": 6,
          "init": undefined,
          "kind": "def_field",
          "name": "b",
          "ref": b: Bool;,
          "type": {
            "id": 5,
            "kind": "type_ref_simple",
            "name": "Bool",
            "optional": false,
            "ref": Bool,
          },
        },
        {
          "args": [],
          "attributes": [],
          "id": 7,
          "kind": "def_function",
          "name": "hello",
          "ref": fun hello() {
        
    },
          "return": null,
          "statements": [],
        },
        {
          "args": [],
          "attributes": [],
          "id": 8,
          "kind": "def_function",
          "name": "hello2",
          "ref": fun hello2() {
        
    },
          "return": null,
          "statements": [],
        },
      ],
      "id": 9,
      "kind": "def_contract",
      "name": "HelloWorld",
      "ref": contract HelloWorld {
    a: Int;
    b: Bool;

    fun hello() {
        
    }

    fun hello2() {
        
    }
},
    },
    "fields": [
      {
        "as": null,
        "default": undefined,
        "index": 0,
        "name": "a",
        "type": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
      },
      {
        "as": null,
        "default": undefined,
        "index": 1,
        "name": "b",
        "type": {
          "kind": "ref",
          "name": "Bool",
          "optional": false,
        },
      },
    ],
    "functions": {
      "hello": {
        "args": [],
        "ast": {
          "args": [],
          "attributes": [],
          "id": 7,
          "kind": "def_function",
          "name": "hello",
          "ref": fun hello() {
        
    },
          "return": null,
          "statements": [],
        },
        "isGetter": false,
        "isMutating": true,
        "isPublic": false,
        "name": "hello",
        "returns": {
          "kind": "void",
        },
        "self": "HelloWorld",
      },
      "hello2": {
        "args": [],
        "ast": {
          "args": [],
          "attributes": [],
          "id": 8,
          "kind": "def_function",
          "name": "hello2",
          "ref": fun hello2() {
        
    },
          "return": null,
          "statements": [],
        },
        "isGetter": false,
        "isMutating": true,
        "isPublic": false,
        "name": "hello2",
        "returns": {
          "kind": "void",
        },
        "self": "HelloWorld",
      },
    },
    "init": null,
    "kind": "contract",
    "name": "HelloWorld",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-3 2`] = `{}`;

exports[`resolveDescriptors should resolve descriptors for case-4 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-4 2`] = `
{
  "sample": {
    "args": [
      {
        "as": null,
        "name": "a",
        "ref": a: Int,
        "type": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
      },
    ],
    "ast": {
      "args": [
        {
          "id": 5,
          "kind": "def_argument",
          "name": "a",
          "ref": a: Int,
          "type": {
            "id": 4,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
        },
      ],
      "attributes": [],
      "id": 6,
      "kind": "def_native_function",
      "name": "sample",
      "nativeName": "hello_world",
      "ref": @name(hello_world)
native sample(a: Int): Int;,
      "return": {
        "id": 3,
        "kind": "type_ref_simple",
        "name": "Int",
        "optional": false,
        "ref": Int,
      },
    },
    "isGetter": false,
    "isMutating": false,
    "isPublic": false,
    "name": "sample",
    "returns": {
      "kind": "ref",
      "name": "Int",
      "optional": false,
    },
    "self": null,
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-5 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {
      "inc": {
        "args": [],
        "ast": {
          "args": [
            {
              "id": 5,
              "kind": "def_argument",
              "name": "self",
              "ref": self: Int,
              "type": {
                "id": 4,
                "kind": "type_ref_simple",
                "name": "Int",
                "optional": false,
                "ref": Int,
              },
            },
          ],
          "attributes": [
            {
              "ref": mutates,
              "type": "mutates",
            },
            {
              "ref": extends,
              "type": "extends",
            },
          ],
          "id": 6,
          "kind": "def_native_function",
          "name": "inc",
          "nativeName": "inc",
          "ref": @name(inc)
mutates extends native inc(self: Int): Int;,
          "return": {
            "id": 3,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
        },
        "isGetter": false,
        "isMutating": true,
        "isPublic": false,
        "name": "inc",
        "returns": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
        "self": "Int",
      },
    },
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-5 2`] = `{}`;

exports[`resolveDescriptors should resolve descriptors for case-6 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {
      "inc": {
        "args": [],
        "ast": {
          "args": [
            {
              "id": 5,
              "kind": "def_argument",
              "name": "self",
              "ref": self: Int,
              "type": {
                "id": 4,
                "kind": "type_ref_simple",
                "name": "Int",
                "optional": false,
                "ref": Int,
              },
            },
          ],
          "attributes": [
            {
              "ref": extends,
              "type": "extends",
            },
          ],
          "id": 10,
          "kind": "def_function",
          "name": "inc",
          "ref": extends fun inc(self: Int): Int {
    return self + 1;
},
          "return": {
            "id": 3,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
          "statements": [
            {
              "expression": {
                "id": 8,
                "kind": "op_binary",
                "left": {
                  "id": 6,
                  "kind": "id",
                  "ref": self,
                  "value": "self",
                },
                "op": "+",
                "ref": self + 1,
                "right": {
                  "id": 7,
                  "kind": "number",
                  "ref": 1,
                  "value": 1n,
                },
              },
              "id": 9,
              "kind": "statement_return",
              "ref": return self + 1;,
            },
          ],
        },
        "isGetter": false,
        "isMutating": false,
        "isPublic": false,
        "name": "inc",
        "returns": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
        "self": "Int",
      },
    },
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-6 2`] = `{}`;

exports[`resolveDescriptors should resolve descriptors for case-7 1`] = `
{
  "Bool": {
    "ast": {
      "id": 2,
      "kind": "primitive",
      "name": "Bool",
      "ref": primitive Bool;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Bool",
    "receivers": {},
  },
  "Int": {
    "ast": {
      "id": 1,
      "kind": "primitive",
      "name": "Int",
      "ref": primitive Int;,
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "primitive",
    "name": "Int",
    "receivers": {},
  },
  "Main": {
    "ast": {
      "declarations": [
        {
          "as": null,
          "id": 5,
          "init": undefined,
          "kind": "def_field",
          "name": "a",
          "ref": a: Int;,
          "type": {
            "id": 4,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
        },
        {
          "args": [
            {
              "id": 7,
              "kind": "def_argument",
              "name": "a",
              "ref": a: Int,
              "type": {
                "id": 6,
                "kind": "type_ref_simple",
                "name": "Int",
                "optional": false,
                "ref": Int,
              },
            },
          ],
          "id": 12,
          "kind": "def_init_function",
          "ref": init(a: Int) {
        self.a = a;
    },
          "statements": [
            {
              "expression": {
                "id": 10,
                "kind": "id",
                "ref": a,
                "value": "a",
              },
              "id": 11,
              "kind": "statement_assign",
              "path": [
                {
                  "id": 8,
                  "kind": "lvalue_ref",
                  "name": "self",
                  "ref": self.a,
                },
                {
                  "id": 9,
                  "kind": "lvalue_ref",
                  "name": "a",
                  "ref": a,
                },
              ],
              "ref": self.a = a;,
            },
          ],
        },
        {
          "args": [],
          "attributes": [
            {
              "ref": get,
              "type": "get",
            },
          ],
          "id": 16,
          "kind": "def_function",
          "name": "hello",
          "ref": get fun hello(): Int {
        return 0;
    },
          "return": {
            "id": 13,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
          "statements": [
            {
              "expression": {
                "id": 14,
                "kind": "number",
                "ref": 0,
                "value": 0n,
              },
              "id": 15,
              "kind": "statement_return",
              "ref": return 0;,
            },
          ],
        },
        {
          "args": [],
          "attributes": [
            {
              "ref": get,
              "type": "get",
            },
          ],
          "id": 20,
          "kind": "def_function",
          "name": "hello2",
          "ref": get fun hello2(): Point {
        return 0;
    },
          "return": {
            "id": 17,
            "kind": "type_ref_simple",
            "name": "Point",
            "optional": false,
            "ref": Point,
          },
          "statements": [
            {
              "expression": {
                "id": 18,
                "kind": "number",
                "ref": 0,
                "value": 0n,
              },
              "id": 19,
              "kind": "statement_return",
              "ref": return 0;,
            },
          ],
        },
      ],
      "id": 21,
      "kind": "def_contract",
      "name": "Main",
      "ref": contract Main {
    a: Int;
    init(a: Int) {
        self.a = a;
    }

    get fun hello(): Int {
        return 0;
    }
    get fun hello2(): Point {
        return 0;
    }
},
    },
    "fields": [
      {
        "as": null,
        "default": undefined,
        "index": 0,
        "name": "a",
        "type": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
      },
    ],
    "functions": {
      "hello": {
        "args": [],
        "ast": {
          "args": [],
          "attributes": [
            {
              "ref": get,
              "type": "get",
            },
          ],
          "id": 16,
          "kind": "def_function",
          "name": "hello",
          "ref": get fun hello(): Int {
        return 0;
    },
          "return": {
            "id": 13,
            "kind": "type_ref_simple",
            "name": "Int",
            "optional": false,
            "ref": Int,
          },
          "statements": [
            {
              "expression": {
                "id": 14,
                "kind": "number",
                "ref": 0,
                "value": 0n,
              },
              "id": 15,
              "kind": "statement_return",
              "ref": return 0;,
            },
          ],
        },
        "isGetter": true,
        "isMutating": true,
        "isPublic": false,
        "name": "hello",
        "returns": {
          "kind": "ref",
          "name": "Int",
          "optional": false,
        },
        "self": "Main",
      },
      "hello2": {
        "args": [],
        "ast": {
          "args": [],
          "attributes": [
            {
              "ref": get,
              "type": "get",
            },
          ],
          "id": 20,
          "kind": "def_function",
          "name": "hello2",
          "ref": get fun hello2(): Point {
        return 0;
    },
          "return": {
            "id": 17,
            "kind": "type_ref_simple",
            "name": "Point",
            "optional": false,
            "ref": Point,
          },
          "statements": [
            {
              "expression": {
                "id": 18,
                "kind": "number",
                "ref": 0,
                "value": 0n,
              },
              "id": 19,
              "kind": "statement_return",
              "ref": return 0;,
            },
          ],
        },
        "isGetter": true,
        "isMutating": true,
        "isPublic": false,
        "name": "hello2",
        "returns": {
          "kind": "ref",
          "name": "Point",
          "optional": false,
        },
        "self": "Main",
      },
    },
    "init": {
      "args": [
        {
          "as": null,
          "name": "a",
          "ref": a: Int,
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
      ],
      "ast": {
        "args": [
          {
            "id": 7,
            "kind": "def_argument",
            "name": "a",
            "ref": a: Int,
            "type": {
              "id": 6,
              "kind": "type_ref_simple",
              "name": "Int",
              "optional": false,
              "ref": Int,
            },
          },
        ],
        "id": 12,
        "kind": "def_init_function",
        "ref": init(a: Int) {
        self.a = a;
    },
        "statements": [
          {
            "expression": {
              "id": 10,
              "kind": "id",
              "ref": a,
              "value": "a",
            },
            "id": 11,
            "kind": "statement_assign",
            "path": [
              {
                "id": 8,
                "kind": "lvalue_ref",
                "name": "self",
                "ref": self.a,
              },
              {
                "id": 9,
                "kind": "lvalue_ref",
                "name": "a",
                "ref": a,
              },
            ],
            "ref": self.a = a;,
          },
        ],
      },
    },
    "kind": "contract",
    "name": "Main",
    "receivers": {},
  },
  "Point": {
    "ast": {
      "fields": [],
      "id": 3,
      "kind": "def_struct",
      "message": false,
      "name": "Point",
      "ref": struct Point {
    
},
    },
    "fields": [],
    "functions": {},
    "init": null,
    "kind": "struct",
    "name": "Point",
    "receivers": {},
  },
}
`;

exports[`resolveDescriptors should resolve descriptors for case-7 2`] = `{}`;
